### PR TITLE
Pcid daemon rework

### PR DESCRIPTION
**Pull request by dsemenets**

### CODE REVIEW

File: tools/libs/light/libxl_pcid.c [hunk 1]
```diff
@@ -94,7 +94,8 @@ static int process_list_assignable(libxl__gc *gc, yajl_gen gen,
     struct dirent *de;
     DIR *dir = NULL;
 
-    dir = opendir(SYSFS_PCIBACK_DRIVER);
+//    dir = opendir(SYSFS_PCIBACK_DRIVER);
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: Do not leave commented out code.
On other hand - good commit description is required.
It is totally unclear why are you doing this change

File: tools/libs/light/libxl_pci.c [hunk 2]
```diff
@@ -754,6 +756,7 @@ bool libxl__is_igd_vga_passthru(libxl__gc *gc,
 
 static int pciback_dev_is_assigned(libxl__gc *gc, libxl_device_pci *pci)
 {
+#if 0
```
Comment (lorc) [inline tools/libs/light/libxl_pci.c]: Don't disable code with preprocessor. Just remove it.

Anyways, git remember everything :)

File: tools/libs/light/libxl_pcid.c [hunk 3]
```diff
@@ -148,7 +148,7 @@ static int pciback_dev_is_assigned(libxl__gc *gc, unsigned int domain,
     if (rc < 0 && errno == ENOENT)
         return 0;
     LOGE(ERROR, "Accessing %s", spath);
-    return -1;
+    return 0;
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c:257]: Why this change is needed?

File: tools/libs/light/libxl_vchan.h [hunk 4]
```diff
@@ -45,6 +45,7 @@ struct vchan_info {
     /* Buffer info. */
     size_t receive_buf_size;
     size_t max_buf_size;
+    int initialized;
```
Comment (lorc) [inline tools/libs/light/libxl_vchan.h]: `bool` will be better in this case

File: tools/include/pcid.h [hunk 5]
```diff
@@ -155,6 +155,34 @@
...
+ * Response (see other mandatory fields above):
+ *  - "resp" field of the response must be set to "is_device_assigned".
+ * Command specific response data:
+ * +-------------+--------------+----------------------------------------------+
+ * | resources   | map          | key 'iomem' - list of memory regions         |
+ * |             |              | key 'irqs' - list of irqs                    |
+ * +-------------+--------------+----------------------------------------------+
+ */
+#define PCID_CMD_RESOURCE_LIST          "resource_list"
+// Arguments
```
Comment (lorc) [inline tools/include/pcid.h]: C++-style comments are prohibited in xen

File: tools/libs/light/libxl_pci.c [hunk 6]
```diff
@@ -81,16 +81,19 @@ static int pci_handle_response(libxl__gc *gc,
     LOG(DEBUG, "command: %s", command_name);
 
     if (strcmp(command_name, PCID_CMD_LIST_ASSIGNABLE) == 0)
-       ret = process_list_assignable(gc, response, result);
+        ret = process_list_assignable(gc, response, result);
```
Comment (lorc) [inline tools/libs/light/libxl_pci.c:84]: Looks like spurious change.

File: tools/libs/light/libxl_pci.c [hunk 7]
```diff
@@ -81,16 +81,19 @@ static int pci_handle_response(libxl__gc *gc,
...
     else if (strcmp(command_name, PCID_CMD_MAKE_ASSIGNABLE) == 0)
         *result = libxl__json_object_alloc(gc, JSON_NULL);
     else if (strcmp(command_name, PCID_CMD_REVERT_ASSIGNABLE) == 0)
         *result = libxl__json_object_alloc(gc, JSON_NULL);
     else if (strcmp(command_name, PCID_CMD_IS_ASSIGNED) == 0)
         *result = (libxl__json_object *)libxl__json_map_get(PCID_MSG_FIELD_RESULT,
-                                                          response, JSON_BOOL);
+                response, JSON_BOOL);
```
Comment (lorc) [inline tools/libs/light/libxl_pci.c:91]: Looks like spurious change.

File: tools/include/pcid.h [hunk 8]
```diff
@@ -155,6 +155,34 @@
...
 #define PCID_MSG_FIELD_RESULT           "result"
 
+/*
+ *******************************************************************************
+ * Get device resources
+ *
+ * This command returns resource list of device
+ *
+ * Request (see other mandatory fields above):
+ *  - "cmd" field of the request must be set to "is_device_assigned".
```
Comment (lorc) [inline tools/include/pcid.h]: is it "is_device_assigned"?

File: tools/include/pcid.h [hunk 9]
```diff
@@ -155,6 +155,34 @@
...
+ *  - "sbdf" SBDF of the device in format defined by PCID_SBDF_FMT.
+ *  - "permissive" pci->permissive field
```
Comment (lorc) [inline tools/include/pcid.h]: What it does exactly? Imagine that someone tries to implement protocol based on this documentation.

File: tools/include/pcid.h [hunk 10]
```diff
@@ -155,6 +155,34 @@
...
+ *
+ * Response (see other mandatory fields above):
+ *  - "resp" field of the response must be set to "is_device_assigned".
```
Comment (lorc) [inline tools/include/pcid.h]: Is it?

File: tools/libs/light/libxl_pcid.c [hunk 11]
```diff
@@ -122,6 +126,127 @@ static int process_list_assignable(libxl__gc *gc, yajl_gen gen,
...
+        make_error_reply(gc, gen, "Can't parse SBDF", command_name);
+        return ERROR_FAIL;
+    }
+
+    json_permissive = libxl__json_map_get(PCID_MSG_FIELD_PERMISSIVE, request, JSON_INTEGER);
+    if (!json_permissive) {
+        make_error_reply(gc, gen, "No mandatory parameter 'permissive'", command_name);
+        return ERROR_FAIL;
+    }
+    permissive = libxl__json_object_get_integer(json_permissive);
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: Should it be boolean?

File: tools/libs/light/libxl_pcid.c [hunk 12]
```diff
@@ -122,6 +126,127 @@ static int process_list_assignable(libxl__gc *gc, yajl_gen gen,
...
+                libxl__json_object_alloc(gc, JSON_INTEGER);
+
+            node->u.i = irq;
+            flexarray_append(irqs->u.array, node);
+    }
+    fclose(f);
+
+    /* Don't restrict writes to the PCI config space from this VM */
+    if (permissive) {
+        if ( sysfs_write_bdf(gc, SYSFS_PCIBACK_DRIVER"/permissive", dom, bus, dev, func)) {
```
Comment (lorc) [inline tools/libs/light/libxl_pcid.c]: Well, this don't look good. "get_resources" operation should return list of resources and don't attempt to change any settings in the pciback.

You need to implement this as a separate operation

**Review by lorc — CHANGES_REQUESTED** [review]

File: tools/libs/light/libxl_pci.c (hunk 2 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pci.c]: Sorry. Corrected

File: tools/libs/light/libxl_pcid.c (hunk 3 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c:257]: This line can be executed only on development stage (perhaps because of selinux or something like this). It should not happen on product. Because of this I didn't implement careful error checking on caller side. So return -1 will be false-positive result on caller side.

File: tools/libs/light/libxl_vchan.h (hunk 4 above)
Comment (dsemenets) [reply tools/libs/light/libxl_vchan.h]: Fixed

File: tools/libs/light/libxl_pcid.c (hunk 1 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c]: fixed

File: tools/libs/light/libxl_pci.c (hunk 6 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pci.c:84]: Align indentations

File: tools/libs/light/libxl_pci.c (hunk 7 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pci.c:91]: Same as previous

File: tools/include/pcid.h (hunk 8 above)
Comment (dsemenets) [reply tools/include/pcid.h]: Corrected

File: tools/include/pcid.h (hunk 9 above)
Comment (dsemenets) [reply tools/include/pcid.h]: Corrected

File: tools/include/pcid.h (hunk 10 above)
Comment (dsemenets) [reply tools/include/pcid.h]: Corrected

File: tools/libs/light/libxl_pcid.c (hunk 11 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c]: Reworked

File: tools/libs/light/libxl_pcid.c (hunk 12 above)
Comment (dsemenets) [reply tools/libs/light/libxl_pcid.c]: Implemented as separate operation
